### PR TITLE
Add support for list, tuple and dict in numeric debugger

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143882

Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D67660049](https://our.internmc.facebook.com/intern/diff/D67660049)

### CODE REVIEW

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 1]
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
     bfs_trace_with_node_process(ep, _assign_debug_handle)
 
 
+def _detach(x):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py]: Add type annotations. `object` is fine if you do isinstance inside it and have a backup if it's not a tensor or a collection of tensors at all.

TypeVar might be helpful for this function

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 2]
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
...
+    elif isinstance(x, dict):
+        detached = {k: _detach(e) for k, e in x.items()}
+    else:
+        detached = x
+    return detached
+
+def _tensor_shape_equals(x, y):
+    if isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor):
+        print("x:", x.shape, " y:", y.shape)
+        print("equals:", x.shape == y.shape)
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py]: No prints committed

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 3]
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
...
+        return x.shape == y.shape
+    elif isinstance(x, (list, tuple)):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py:90]: check that y is a list or tuple too

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 4]
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
...
+        detached = x.detach()
+    elif isinstance(x, (list, tuple)):
+        detached = type(x)([_detach(e) for e in x])
+    elif isinstance(x, dict):
+        detached = {k: _detach(e) for k, e in x.items()}
+    else:
+        detached = x
+    return detached
+
+def _tensor_shape_equals(x, y):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py]: For these two functions, see if you can reuse `map_aggregate` from fx.node. It already handles all the tuple/dict branching stuff.
You can give it a callable for Tensor to do the detach or replace a Tensor with its shape for comparison

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 5]
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
...
+    elif isinstance(x, dict) and isinstance(y, dict):
+        all_equal = True
+        for k in x:
+            all_equal = all_equal and k in y and (_tensor_shape_equals(x[k], y[k]))
+        return all_equal
+    else:
+        print(f"Comparing non Tensors: {x} and {y}, they must be equal")
+        return type(x) == type(y) and x == y
+
+def _loss_fn(loss, x, y):
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py]: Same here, reuse `map_aggregate` for this.
Also, type annotate `loss`

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 6]
```diff
@@ -90,11 +129,10 @@ def __init__(
         self.node_name = node_name
         self.nn_module_stack = nn_module_stack
         self.debug_handle = debug_handle
-        self.stats: List[torch.Tensor] = []
+        self.stats: List[Any] = []
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py:140]: Use `object` instead of `Any`

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 7]
```diff
@@ -163,26 +201,16 @@ class QuantizationComparisonResult:
...
         self, loss_function: Callable[[torch.Tensor, torch.Tensor], torch.Tensor]
     ) -> torch.Tensor:
-        if self.actual.shape != self.ref.shape:
-            raise ValueError(
-                f"Cannot compare tensors with different shapes: {self.actual.shape} vs {self.ref.shape}"
-            )
-        return loss_function(
-            self.actual.to(dtype=torch.float32), self.ref.to(dtype=torch.float32)
-        )
+        return _loss_fn(loss_function, self.actual, self.ref)
```
Comment (dulinriley) [inline torch/ao/quantization/pt2e/_numeric_debugger.py:221]: You might want to leave in the `to(torch.float32)`, as most loss functions require floats.

Otherwise `self.loss(F.mse_loss)` fails for integer tensors

File: torch/testing/_internal/common_quantization.py [hunk 8]
```diff
@@ -2930,6 +2930,21 @@ def forward(self, x1, x2, x3, x4):
...
 
+    class Conv2dWithSplit(torch.nn.Module):
+        def __init__(self) -> None:
+            super().__init__()
+            self.conv1 = torch.nn.Conv2d(3, 3, 3)
+            self.conv2 = torch.nn.Conv2d(3, 3, 3)
+
+        def forward(self, x):
+            x = self.conv1(x)
+            x1, x2 = x.split(2, dim=1)
```
Comment (dulinriley) [inline torch/testing/_internal/common_quantization.py:2941]: Can you add a comment explaining why `split` is important? Is it because it returns a list/tuple of tensors?

**Review by dulinriley — APPROVED** [review]
Thanks for adding this functionality

File: torch/ao/quantization/pt2e/_numeric_debugger.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/_numeric_debugger.py]: I'm going to use Any

File: torch/ao/quantization/pt2e/_numeric_debugger.py (hunk 4 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/_numeric_debugger.py]: I can use map_aggregate for `_detach`, but others does not help that much I think

File: torch/testing/_internal/common_quantization.py (hunk 8 above)
Comment (jerryzh168) [reply torch/testing/_internal/common_quantization.py:2941]: yes

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).

File: torch/ao/quantization/pt2e/_numeric_debugger.py (hunk 4 above)
Comment (jerryzh168) [reply torch/ao/quantization/pt2e/_numeric_debugger.py]: actually I'll revert this since it introduced a lot of typing errors..

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).
